pkg/operator/controller/dns_status.go

```diff
@@ -63,39 +77,98 @@ func computeDNSStatusConditions(dns *operatorv1.DNS, clusterIP string, haveDNSDa
+	DNSNoDNSPodsAvailable = "NoDNSPodsAvailable"
+
+	// DNSInvalidDNSMaxUnavailable indicates that the DNS daemonset has an invalid
+	// MaxUnavailable value configured.
+	DNSInvalidDNSMaxUnavailable = "InvalidDNSMaxUnavailable"
+
+	// DNSMaxUnavailableDNSPodsExceeded indicates that the number of unavailable DNS
+	// pods is greater than the configured MaxUnavailable.
+	DNSMaxUnavailableDNSPodsExceeded = "MaxUnavailableDNSPodsExceeded"
+)
+
 // computeDNSDegradedCondition computes the dns Degraded status
 // condition based on the status of clusterIP and the DNS daemonset.
 // The node-resolver daemonset is not a part of the calculation of
 // degraded condition. If the elapsed time between currentTime and
 // oldCondition.LastTransitionTime is <= transitionUnchangedToleration
 // then consider oldCondition to be recent and return oldCondition to
 // prevent frequent updates.
-func computeDNSDegradedCondition(oldCondition *operatorv1.OperatorCondition, clusterIP string, haveDNSDaemonset bool, dnsDaemonset *appsv1.DaemonSet, transitionUnchangedToleration time.Duration, currentTime time.Time) operatorv1.OperatorCondition {
-	degradedCondition := &operatorv1.OperatorCondition{
-		Type: operatorv1.OperatorStatusTypeDegraded,
+func computeDNSDegradedCondition(oldDegradedCondition, newProgressingCondition *operatorv1.OperatorCondition, clusterIP string, haveDNSDaemonset bool, dnsDaemonset *appsv1.DaemonSet, transitionUnchangedToleration time.Duration, currentTime time.Time) (operatorv1.OperatorCondition, error) {
+	degradedCondition := operatorv1.OperatorCondition{
+		Type:   operatorv1.OperatorStatusTypeDegraded,
+		Status: operatorv1.ConditionUnknown,
 	}
-	status := operatorv1.ConditionUnknown
-	degradedReasons := []string{}
-	messages := []string{}
+	graceCondition := operatorv1.OperatorCondition{
+		Status: operatorv1.ConditionUnknown,
+	}
+	var currentConditions []operatorv1.OperatorCondition
+	var degradedConditions []operatorv1.OperatorCondition
+
+	transitionTime := metav1.NewTime(currentTime)
+
+	// The list of status conditions we might see when the operator is NOT degraded.
+	expected := []cond.ExpectedCondition{
+		{
+			Condition: DNSNoService,
+			Status:    operatorv1.ConditionFalse,
+		},
+		{
+			Condition: DNSNoDNSDaemonSet,
+			Status:    operatorv1.ConditionFalse,
+		},
+		{
+			Condition: DNSNoDNSPodsDesired,
+			Status:    operatorv1.ConditionFalse,
+		},
+		{
+			Condition: DNSNoDNSPodsAvailable,
+			Status:    operatorv1.ConditionFalse,
+		},
+		{
+			Condition: DNSInvalidDNSMaxUnavailable,
+			Status:    operatorv1.ConditionFalse,
+		},
+		{
+			Condition: DNSMaxUnavailableDNSPodsExceeded,
+			Status:    operatorv1.ConditionFalse,
+			// This grace period is the length of time permitted for a new status before changing a previous status.
+			// In other words, we don't change status from false to true until the grace period has passed.
+			GracePeriod: transitionUnchangedToleration,
```

Comment:
I believe this there is a logic error here. Once `GracePeriod` is added, this logic will always result in adding the condition to `graceConditions` (referring to [checkCondition](https://github.com/openshift/cluster-dns-operator/pull/373/files#diff-32495132facf7e0819a407af732514958b198d9e40236656bc43b093345d1539R209) logic), it will never be  considered `degraded`. 

That's because of we always pass the current time (`now`) into `computeDNSDegradedCondition`:
```
degradedCondition, err := computeDNSDegradedCondition(oldDegradedCondition, &newProgressingCondition, clusterIP, haveDNSDaemonset, dnsDaemonset, transitionUnchangedToleration, now)
```

These new itemized conditions are stateless. By that, I mean they are generated each time we enter `computeDNSDegradedCondition`, and they are not actually embedded into the DNS Operator's status. This means their last transition time (`transitionTime` var as you called it), is _always_ the current time. We don't store their history or transition time anywhere.

If the transition time is always the current time (now) for these conditions, `checkConditions` will never consider the transition time (now) after the current time (now) regardless of how long the graceCondition is. 

The cluster-ingress-operator doesn't have this issue because it stores these itemized conditions in the IngressController object (presenting the itemized status conditions to the end user). 

I'm not sure if there is an easy fix other than converting the itemized conditions (like `DNSMaxUnavailableDNSPodsExceeded`) to be embedded into the DNS Operator object so we track their transition time. Otherwise, adding `DNSMaxUnavailableDNSPodsExceeded` in `ExpectedCondition` does nothing, and same with the [grace period logic](https://github.com/openshift/cluster-dns-operator/pull/373/files#diff-32495132facf7e0819a407af732514958b198d9e40236656bc43b093345d1539R226-R233) for `checkConditions`.

Reply:
> This means their last transition time (transitionTime var as you called it), is always the current time. We don't store their history or transition time anywhere.

I expected that in `mergeConditions` that the `transitionTime` wouldn't be updated if there were [no changes to the condition](https://github.com/openshift/cluster-dns-operator/pull/373/files#diff-69f15ff9fa97586b9285d57454f58562613c7c19230815edbb3362d0b18be44aR39).  I'll need to check into whether that is working as expected.

Reply:
I'll need to replace the existing [`mergeConditions`](https://github.com/openshift/cluster-dns-operator/blob/master/pkg/operator/controller/status/controller.go#L620)